### PR TITLE
[GSoC] physics/continuum_mechanics: Added the `draw` method for the Truss Class

### LINKED ISSUE #24040: Truss - create and expose an attribute to get the length of the members
The `Truss` class can be used as a starting point to compute node deflections thanks to the principle of virtual works or energy methods. However, these computations require the knowledge of the member lengths. The class already computes them internally. It would be nice to be able to access them via some attribute: `members_lengths` maybe?

### PR DESCRIPTION
**Pull request by AdvaitPote**
#### References to other Issues or PRs

Fixes #24040 

#### Brief description of what is fixed or changed
As a final step for the `Truss` class, the `draw` method for the same has been added. This method, similar to the one in `Beam`, returns a plot object representing the state of the truss denoting all of its properties like loads and supports along with its structural components like nodes and members. 

#### Release Notes

* physics.continuum_mechanics
  * Added the `draw` method for getting the diagram of a given `truss`.

### CODE REVIEW

**Comment by AdvaitPote** [comment]
@Psycho-Pirate

File: sympy/physics/continuum_mechanics/truss.py [hunk 1]
```diff
@@ -733,3 +737,226 @@ def solve(self):
...
+            The user must be careful while entering load values their
+            directions. The draw function assumes a sign convention which
+            is used for plotting loads.
+
+            Given a right handed coordinate system with XYZ coordinates,
+            the supports are assumed to be such that reaction forces of a
+            pinned support are in the +X and +Y direction while those of a
+            roller support are in the +Y direction. For the load, the range
+            of angles one can input goes all the way to 360 degrees which, in the
+            plot, is the angle that the load vector makes with positive x-axis.
```
Comment (Psycho-Pirate) [inline sympy/physics/continuum_mechanics/truss.py:757]: Just add the direction of the angle as well after "positive x-axis"

**Comment by Psycho-Pirate** [comment]
Its working fine. I was thinking if we could make it a bit more appealing. Maybe add some thickness to the members. We can discuss this.

**Comment by AdvaitPote** [comment]
I have added the changes recommended. The members have been replaced by rectangles rotated by an angle determined by the coordinates of the two nodes of the member. Also, I was thinking the doctests for the method can be a little more developed so I have replaced them with the example you had mentioned above. This truss I feel because of its symmetricity and because it is a bit more complicated, makes documentation a little better.

```
>>> from sympy.physics.continuum_mechanics.truss import Truss
>>> t = Truss()
>>> t.add_node("A", -4, 0)
>>> t.add_node("B", 0, 0)
>>> t.add_node("C", 4, 0)
>>> t.add_node("D", 8, 0)
>>> t.add_node("E", 6, 2/math.sqrt(3))
>>> t.add_node("F", 2, 2*math.sqrt(3))
>>> t.add_node("G", -2, 2/math.sqrt(3))
>>> t.add_member("AB","A","B")
>>> t.add_member("BC","B","C")
>>> t.add_member("CD","C","D")
>>> t.add_member("AG","A","G")
>>> t.add_member("GB","G","B")
>>> t.add_member("GF","G","F")
>>> t.add_member("BF","B","F")
>>> t.add_member("FC","F","C")
>>> t.add_member("CE","C","E")
>>> t.add_member("FE","F","E")
>>> t.add_member("DE","D","E")
>>> t.apply_support("A","pinned")
>>> t.apply_support("D","roller")
>>> t.apply_load("G", 3, 90)
>>> t.apply_load("E", 3, 90)
>>> t.apply_load("F", 2
>>> p = t.draw()
>>> p.show()
```
Screenshot 2022-09-05 at 10 55 34 AM)

**Comment by Psycho-Pirate** [comment]
> I have added the changes recommended. The members have been replaced by rectangles rotated by an angle determined by the coordinates of the two nodes of the member. Also, I was thinking the doctests for the method can be a little more developed so I have replaced them with the example you had mentioned above. This truss I feel because of its symmetricity and because it is a bit more complicated, makes documentation a little better.

This is looking really good. The docs are looking great as well. I was thinking that maybe we can replace the initial example truss image with the one created by the draw function as well.

Also one small nitpick, some nodes seem a bit off. I tried zooming in on nodes B and C. 
For node B it shows 
nb

For node C
nc

It looks like node B is a bit off.
Similarly, nodes D, E and F seem a bit off too. Just look into it.

**Comment by Davide-sd** [comment]
@AdvaitPote very nice work on the `Truss` class.

In the last few days I have been trying it on different problems, specifically the ones dealing with the principal of virtual works and energy methods in order to compute the displacements of the nodes.

Sometime it is useful to provide coordinates with physical units. The symbolic part of the code dealing with internal forces computation is going to work well. However, physical units on coordinates will make the `draw` method fails, because coordinates might contain `Symbol` or `Quantity` objects, which are not numerical data. Nonetheless, visualizing a frame is extremely important to verify the user inputs.

So, I'm going to post an example of a calculation involving the `Truss` class:

```python
from sympy import *
from sympy.physics.units import m, mm, N
from sympy.physics.continuum_mechanics import Truss

def mme(*args):
    """Return a element-by-element multiplication of SymPy matrices.
    """
    if len(args) < 1:
        raise ValueError("Please, provide at least two matrices.")
    r = args[0]
    for i in range(1, len(args)):
        r = matrix_multiply_elementwise(r, args[i])
    return r

def get_lengths(t):
    """Compute the length of the members of a Truss"""
    members_lengths = dict()
    for member in list(t._members):
        start, end = t._members[member]
        members_lengths[member] = sqrt((t._node_coordinates[start][0]-t._node_coordinates[end][0])**2 + (t._node_coordinates[start][1]-t._node_coordinates[end][1])**2)
    return members_lengths

# length
L = 4 * m

def build_structure():
    t = Truss()
    t.add_node("A", 0, 0)
    t.add_node("B", L, 0)
    t.add_node("C", 2 * L, 0)
    t.add_node("D", 3 * L, 0)
    t.add_node("E", L, L)
    t.add_node("F", 2 * L, L)
    t.add_member("AB", "A", "B")
    t.add_member("BC", "B", "C")
    t.add_member("CD", "C", "D")
    t.add_member("AE", "A", "E")
    t.add_member("BE", "B", "E")
    t.add_member("BF", "B", "F")
    t.add_member("CF", "C", "F")
    t.add_member("DF", "D", "F")
    t.add_member("EF", "E", "F")
    t.apply_support("A", type="pinned")
    t.apply_support("D", type="roller")
    return t

# solve for actual forces
t1 = build_structure()
t1.apply_load("E", magnitude=40e03 * N, direction=270)
t1.apply_load("C", magnitude=100e03 * N, direction=270)
t1.solve()

# NOTE: this fails because coordinates contains a `Quantity`
# the same happens if they contain symbols...
# t1.draw().show()

# unit load
t2 = build_structure()
t2.apply_load("B", magnitude=1, direction=270)
t2.solve()

# cross sectional area
A = 1800 * mm**2
E = 200e03 * N / mm**2

L = Matrix(list(get_lengths(t1).values()))
# actual forces
F = Matrix(list(t1.internal_forces.values()))
# virtual forces
Fv = Matrix(list(t2.internal_forces.values()))

# principle of virtual work
displ = sum(mme(F, Fv, L) / (A * E)).n()
displ
# out: 0.00351524771279376m
```

As noted in the code, drawing this structure will raise an error. I propose the following:
* modify the signature of `draw(self)` to `draw(self, subs_dict=None)`, where `subs_dict` is a user-provided substitution dictionary. In the above example:  `t1.draw(subs_dict={m: 1})`. Here, the user is assigning a value of 1 to a meter, allowing the method to draw the structure.
* inside the `draw` method, extract every free symbols and `Quantity` from the nodes coordinates. Implement a (set) difference between these symbolic objects and the keys of `subs_dict`. If the length of the resulting set is greater than zero, raise a `ValueError` and inform the user that in order to proceed, a substitution dictionary is required.

**Comment by Psycho-Pirate** [comment]
Thanks for pointing this out! @Davide-sd

**Comment by AdvaitPote** [comment]
I have added the `member_lengths` attribute as was requested in the issue #24040 pointed out by @Davide-sd. Also, I will be adding additional examples and tests including only Symbolics in the future commits. 

I am looking into the issues pointed out above. I have my mid-semester examinations going on so it might take some time.

**Comment by AdvaitPote** [comment]
I have added `subs_dict` to the method.

**Comment by AdvaitPote** [comment]
@Psycho-Pirate I have made the required changes. The node misalignment was arising from the way the starting point of each member rectangle `xy` is calibrated. As per various possible cases of the initial node and the final node, modifications were required to be made to the `xy`, `width`, and `height` attributes of the member. The nodes look better now.

This is node B now. 
Screenshot 2022-10-12 at 11 43 02 PM

And this is node C. 
Screenshot 2022-10-13 at 1 04 49 AM

**Comment by AdvaitPote** [comment]
@Psycho-Pirate

**Comment by oscarbenjamin** [comment]
Closing and reopening to rerun the tests.

**Comment by oscarbenjamin** [comment]
@Psycho-Pirate does this look good to you?

**Comment by Psycho-Pirate** [comment]
Sorry for such a delayed response! This looks good to me! 
Please restart the tests @oscarbenjamin

**Comment by Psycho-Pirate** [comment]
The mailmap file that is present in your branch doesn't contain some names. These names are present in the master branch. I think you should just replace your mailmap with the master's mailmap and it should pass.
@AdvaitPote

**Comment by Psycho-Pirate** [comment]
It looks like your master isn't upto date with Sympy's master

**Comment by AdvaitPote** [comment]
Resolved! @Psycho-Pirate

**Pull request merged**